106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,518 @@
+The `StrimziCruiseControlNotifier` will be kept in a separate repository and will override all the methods declared by the `AnomalyNotifier` interface.
+We will provide a concrete implementations of the `AnomalyNotifier.alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create (if it doesn't already exist) or update a ConfigMap with the name `<cluster-name>-goal-violation-map`.
+This ConfigMap will contain all the detected anomalies for a given Kafka cluster.
+
+The ConfigMap will look like this:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: my-cluster-goal-violation-map
+data:
+  fixableAnomalies: |
+    - <anomaly-id-1>:
+        - <fixable-goal-1>
+        - <fixable-goal-2>
+    - <anomaly-id-2>:
+        - <fixable-goal-3> 
+# ...
+```
+
+The advantages of using a single ConfigMap for all the detected anomalies versus one per anomaly are:
+1. It provides a single point of reference for all the anomalies
+2. When dealing with multiple Kafka clusters, the number of ConfigMap(s) created will be equal to the number of clusters deployed
+
+This generated ConfigMap will contain the list of fixable goal violation anomalies in the section named `fixableAnomalies`.
+The `fixableAnomalies` list will state the `anomalyId` of the anomaly as well as the fixable violated goals.
+The `fixableAnomalies` list will be populated by the notifier every time it detects an anomaly.
+Whenever a rebalance is about to be triggered by the operator the `fixableAnomalies` list will be cleared.
+This will be the case for manual rebalances triggered by the user, as well as auto-rebalance rebalances that are happening in `add-broker`, `remove-broker`, or `imbalance` mode.
+This is because not matter the reason for triggering the rebalance it is possible that the rebalance will address some or all of the detected anomalies.
+If some other anomalies are detected again after the rebalance then they will be updated in the ConfigMap by the notifier and an `imbalance` related rebalance will take place.
+
+The ConfigMap will also have an optional field called `unfixableAnomalies` list which would be created only when an unfixable anomaly is detected i.e. an anomaly having unfixable goals.
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: my-cluster-goal-violation-map
+data:
+  fixableAnomalies: |
+    - <anomaly-id-1>:
+        - <fixable-goal-1>
+        - <fixable-goal-2>
+    - <anomaly-id-2>:
+        - <fixable-goal-3> 
+  unfixableAnomalies: |
+    - <anomaly-id-3>:
+        - <unfixed-goal-1>
+        - <unfixed-goal-2>
+    - <anomaly-id-4>:
+        - <unfixed-goal-3> 
```

Comment:
We are strictly separating `fixableAnomalies` from `unfixableAnomalies` so fixable vs unfixable goals but I have a question ... is it possible to have a fixable anomaly which brings some fixable goals but other unfixable goals? Because in such a case, the above schema would not work.

Reply:
if an Anomaly has both fixable and unfixable goals, then I was planning to have the anomaly reflected in both the lists `fixableAnomalies` and `unfixableAnomalies`